### PR TITLE
feat: AIP-123 – Complain if resource references are not strings.

### PR DESCRIPTION
**Pull request by lukesneeringer**
/cc @jskeet

### CODE REVIEW

File: docs/rules/0123/resource-reference-type.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+  Author author = 2 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Author"
+  }];
+}
+```
+
+**Correct** code for this rule:
+
+```proto
+// Incorrect.
```
Comment (apasel422) [inline docs/rules/0123/resource-reference-type.md:45]:
```suggestion
// Correct.
```

File: docs/rules/0123/resource-reference-type.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+message Book {
+  string name = 1;
+
+  string author = 2 [(google.api.resource_reference) = {
+    type: "library.googleapis.com/Author"
+  }];
+}
```
Comment (apasel422) [inline docs/rules/0123/resource-reference-type.md:61]: Remove this second `message Book`.

File: rules/aip0123/resource_reference_type_test.go [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+	annotation := ` [(google.api.resource_reference) = {
+		type: "library.googleapis.com/Author"
+	}]`
+	for _, test := range []struct {
+		name       string
+		Type       string
+		Annotation string
+		problems   testutils.Problems
+	}{
+		{"Valid", "string", annotation, nil},
```
Comment (apasel422) [inline rules/aip0123/resource_reference_type_test.go:33]: Add a valid case for `repeated string` as well.

File: rules/aip0123/resource_reference_type_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+			  import "google/api/resource.proto";
+				message Book {
+					{{.Type}} author = 1{{.Annotation}};
+				}
+				message Author {}  // For the failure case, to make compilation work.
+			`, test)
+			field := file.GetMessageTypes()[0].GetFields()[0]
+			got := resourceReferenceType.Lint(file)
+			if diff := test.problems.SetDescriptor(field).Diff(got); diff != "" {
+				t.Errorf(diff)
```
Comment (apasel422) [inline rules/aip0123/resource_reference_type_test.go:48]:
```suggestion
				t.Error(diff)
```

**Review by apasel422 — APPROVED** [review]

File: docs/rules/0123/resource-reference-type.md (hunk 2 above)
Comment (lukesneeringer) [reply docs/rules/0123/resource-reference-type.md:61]: No, this one is indeed correct, and a valid thing that someone might want to do (even if it is not the most likely problem someone is hitting if they break this rule).

File: rules/aip0123/resource_reference_type_test.go (hunk 3 above)
Comment (lukesneeringer) [reply rules/aip0123/resource_reference_type_test.go:33]: Done.

File: docs/rules/0123/resource-reference-type.md (hunk 2 above)
Comment (apasel422) [reply docs/rules/0123/resource-reference-type.md:61]: In that case, shouldn't the comment above the message be `// Correct`?

**Pull request merged**